CL2-6612 Smarter smart groups

sebastienhoorens opened the pull request.

sebastienhoorens committed:
pair programming smarter smart groups

sebastienhoorens committed:
fixed spec issue with validating rules by schema

sebastienhoorens committed:
rule descriptions but migrating spec locales in progress

sebastienhoorens committed:
loading spec translations

sebastienhoorens committed:
minor

jinjagit committed:
Add multivalue predicates to smart group topic rules

jinjagit committed:
Edit English translations for Smart Groups Topic rules

jinjagit committed:
Move and edit translations to spec/fixtures/locales/en.yml

Based on changes in:
https://github.com/CitizenLabDotCo/citizenlab/pull/545

jinjagit committed:
Move and edit translations to spec/fixtures/locales/fr-FR.yml

jinjagit committed:
Move and edit translations to spec/fixtures/locales/nl-NL.yml

jinjagit committed:
Only change translations for smart groups topics, not ideas.

Use original translations of smart groups ideas, so that we can
focus on the topic related specs only.

jinjagit committed:
Add spec_helper.rb for Smart groups engine

jinjagit committed:
Return arrays of Topics from mulitvalue predicates + edit specs

Need to use Topic.where...to_a, not [Topic.where...] to return
an array of Topics, not a collection of ActiveRecord::Relations

jinjagit committed:
Rename var

jinjagit committed:
Delete whitespace

sebastienhoorens committed:
Merge branch 'master' into CL2-6612-smarter-smart-groups

jinjagit committed:
Don't pass array for single-predicates, rather use a collection

jinjagit committed:
Don't use ids in with_some_topics scope in Idea model

Insead use topics collection passed in

jinjagit committed:
Merge branch 'CL2-6612-smarter-smart-groups' into CL2-6691-smarter-smart-groups

jinjagit committed:
Merge pull request #576 from CitizenLabDotCo/CL2-6691-smarter-smart-groups

CL2-6691 Smarter Smart groups:  Add multi-value predicates to smart group topic rules (back-end)

jinjagit committed:
Add multi-value predicates for projects + related specs

Now have 1 failing test in:
back/engines/commercial/smart_groups/spec/services/side_fx_project_spec.rb

jinjagit commented:
After adding multi-value predicates for projects + related specs, etc., we now have one failing test in `back/engines/commercial/smart_groups/spec/services/side_fx_project_spec.rb`
```
Failures:

  1) SideFxProjectService before_destroy destroys any smart group that refers to this project
     Failure/Error:
       group1 = create(:smart_group, rules: [{
         ruleType: 'participated_in_project',
         predicate: 'in',
         value: project.id
       }])
     
     ActiveRecord::RecordInvalid:
       Validation failed: Rules {:fragment=>"#/0", :error=>"AnyOf", :human_message=>"The property '#/0' of type object did not match one or more of the required schemas in schema 03301f12-482e-5f27-bf58-79417cb55f75"}
     # /usr/local/bundle/gems/factory_bot-6.2.0/lib/factory_bot/evaluation.rb:18:in `create'
     # /usr/local/bundle/gems/factory_bot-6.2.0/lib/factory_bot/strategy/create.rb:12:in `block in result'
     # /usr/local/bundle/gems/factory_bot-6.2.0/lib/factory_bot/strategy/create.rb:9:in `tap'
     # /usr/local/bundle/gems/factory_bot-6.2.0/lib/factory_bot/strategy/create.rb:9:in `result'
     # /usr/local/bundle/gems/factory_bot-6.2.0/lib/factory_bot/factory.rb:43:in `run'
     # /usr/local/bundle/gems/factory_bot-6.2.0/lib/factory_bot/factory_runner.rb:29:in `block in run'
     # /usr/local/bundle/gems/factory_bot-6.2.0/lib/factory_bot/factory_runner.rb:28:in `run'
     # /usr/local/bundle/gems/factory_bot-6.2.0/lib/factory_bot/strategy_syntax_method_registrar.rb:28:in `block in define_singular_strategy_method'
     # ./engines/commercial/smart_groups/spec/services/side_fx_project_spec.rb:11:in `block (3 levels) in <top (required)>'
     # /usr/local/bundle/gems/webmock-3.12.1/lib/webmock/rspec.rb:37:in `block (2 levels) in <top (required)>'
```
I have not had time to look into this yet.

sebastienhoorens committed:
Merge branch 'master' into CL2-6612-smarter-smart-groups

sebastienhoorens committed:
no more need to load translation in each smart group test

sebastienhoorens committed:
fixed issues with deleting referencing groups and loading main spec translations

sebastienhoorens committed:
fixed user spec

sebastienhoorens committed:
check if rule is multivalue and refactor ideas with topics calls to pass topics instead of ids

sebastienhoorens committed:
small fix

sebastienhoorens committed:
one query for smart group rules filter

sebastienhoorens committed:
validate value after all

sebastienhoorens committed:
small fixes: subset and multivalue pred enum

sebastienhoorens committed:
smart group multi projects frontend

sebastienhoorens committed:
frontend support for multivalue smartgroup by topic or idea status participation

sebastienhoorens committed:
all participation rules applied in one query

sebastienhoorens committed:
project participants in one query

sebastienhoorens committed:
updated changelog

sergei-kucher commented on back/Gemfile:
```diff
@@ -38,13 +38,15 @@ group :development, :test do
   gem 'rspec_junit_formatter'
   gem 'rspec-rails'
   gem 'rspec-parameterized'
+  gem 'rspec-sqlimit'
```
👍

brentguf committed:
IdeaStatusValueSelector: interface instead of type

brentguf committed:
IdeaStatusValueSelector: refactor to functional component

brentguf committed:
IdeaStatusValueSelector: resource hook instead of resource component

brentguf committed:
IdeaStatusValueSelector: useLocalize

brentguf committed:
IdeaStatusValueSelector: more detailed prop

brentguf committed:
IdeaStatusValueSelector: export props (build error)

brentguf commented on front/app/modules/commercial/smart_groups/components/UserFilterConditions/rules.ts:
```diff
@@ -487,43 +515,43 @@ export const ruleTypeConstraints = {
   },
   participated_in_project: {
-    in: ProjectValueSelector,
+    in: ProjectValuesSelector,
```
Is it desired behavior that for the positive selections there is a multi select, while for negative selections (e.g. not_posted_in, not_commened_in, etc.) only 1 value can be selected?

sebastienhoorens replied:
Yes, that's the requirement. For negation, multiple values do not make much sense semantically so we keep single values in those cases.

jinjagit replied:
The negatives are definitely not included in the `MULTIVALUE_PREDICATES`.

I seem to remember us discussing this, and I _think_ we felt that needing to select members of a smart group based on a predicate of _not_ having done one of 'x OR y OR ...' was unlikely (though I can kind of imagine use cases), but I could be mis-remembering. 

Maybe we can wait and see if there is a demand, and if so we add this functionality?

sebastienhoorens replied:
There would be two ways to interpret negation on multiple values:
- No participation in any of those projects. This is currently already possible by adding multiple rules (one for each value).
- You didn't participate in one of those projects. Could be useful, but I think there will be a very low demand for this. Currently the work around for this is to create multiple groups (one for each value).

brentguf reviewed (commented):
@sebastienhoorens @jinjagit See my question below.

The front-end components are okay, but they are not written in the way we do it now. Simon, I refactored one as an example. Would you perhaps like to have a go at the other two to get practice with it? I've made some clear commits with the steps I took. Note that the component can probably still be improved (e.g. I'm not sure memo, which replaces PureComponent, has benefits here).

Otherwise, if you're short on time and my question is answered, you can merge like this.

jinjagit commented:
> The front-end components are okay, but they are not written in the way we do it now. Simon, I refactored one as an example. Would you perhaps like to have a go at the other two to get practice with it? ...

@brentguf @sebastienhoorens  I'd love to have a go! :+1:

brentguf commented:
> I'd love to have a go! 👍

@jinjagit don't hesitate to reach out if you have questions about why I did something.

jinjagit committed:
ProjectValueSelector: interface, not type

jinjagit committed:
ProjectValueSelector: refactor to functional component

jinjagit commented:
@brentguf In the 3rd of your commit steps, “resource hook instead of resource component”…  
  
What is the equivalent of `import useIdeaStatuses from 'hooks/useIdeaStatuses';` for the Projects version?
`import useProjects from hooks/use<???>;`
A `useProjects` (plural) hook doesn’t seem to exist (a singular `useProject` hook does exist, however).  
  
Also, `UserFilterConditions/ValueSelector/ProjectValueSelector.tsx` also uses `projectsList`, for example in:
```
return projects.projectsList.map((project) => { ... }
```
but I can see no matching `useProjectsList` hook either.  
  
How should these differences be resolved?

brentguf commented:
> @brentguf In the 3rd of your commit steps, “resource hook instead of resource component”…
> 
> What is the equivalent of import useIdeaStatuses from 'hooks/useIdeaStatuses'; for the Projects version?
> import useProjects from hooks/use<???>;
> A useProjects (plural) hook doesn’t seem to exist (a singular useProject hook does exist, however).

@jinjagit Yeah, it doesn't exist yet. I think it's challenging to make it as your first hook (but if you want to try, go for it!). So for this component, I'd advise to keep using GetProjects for now.

> Also, UserFilterConditions/ValueSelector/ProjectValueSelector.tsx also uses projectsList, for example in:
> 
> return projects.projectsList.map((project) => { ... }
> but I can see no matching useProjectsList hook either.
> 
> How should these differences be resolved?

projectsList comes from projects (from the data you get back when using GetProjects). It's a bit confusing.

jinjagit commented:
@brentguf So, sticking with `GetProjects`, I am guessing I cannot just replace
```
const ProjectValuesSelectorWithHOC = localize(ProjectValuesSelector);

export default (inputProps) => (
  <GetProjects publicationStatuses={['draft', 'published', 'archived']}>
    {(projects) => (
      <ProjectValuesSelectorWithHOC {...inputProps} projects={projects} />
    )}
  </GetProjects>
);
```
(at the end of the file), with
```
export default ProjectValueSelector;
```
... as in your example. I think I need to involve `GetProjects` in this change? How to do this?

brentguf commented:
> @brentguf So, sticking with `GetProjects`, I am guessing I cannot just replace
> 
> ```
> const ProjectValueSelectorWithHOC = localize(ProjectValueSelector);
> 
> export default (inputProps: Props) => (
>   <GetProjects publicationStatuses={['draft', 'published', 'archived']}>
>     {(projects) => (
>       <ProjectValueSelectorWithHOC {...inputProps} projects={projects} />
>     )}
>   </GetProjects>
> );
> ```
> 
> (at the end of the file), with
> 
> ```
> export default ProjectValueSelector;
> ```
> 
> ... as in your example. I think I need to involve `GetProjects` in this change? How to do this?

I'm not sure what you mean. You can just leave the GetProjects part intact for this component then and it also does the export. You have to look at it as a component that fetches data and passes it on to the component it's wrapping (ProjectValueSelectorWithHOC). 

If you're not using the localize wrapper anymore, you can replace ProjectValueSelectorWithHOC inside this GetProjects with ProjectValueSelector.

jinjagit commented:
> I'm not sure what you mean. You can just leave the GetProjects part intact for this component then and it also does the export. You have to look at it as a component that fetches data and passes it on to the component it's wrapping (ProjectValueSelectorWithHOC).

@brentguf Sorry. I think I confused the `Value` and `Values` files.

jinjagit committed:
ProjectValueSelector: useLocalize

jinjagit committed:
ProjectValueSelector: more detailed prop

jinjagit committed:
ProjectValueSelector: export props (build error)

brentguf commented:
> > I'm not sure what you mean. You can just leave the GetProjects part intact for this component then and it also does the export. You have to look at it as a component that fetches data and passes it on to the component it's wrapping (ProjectValueSelectorWithHOC).
> 
> @brentguf Sorry. I think I confused the `Value` and `Values` files.

Not following, but let me know if you're stuck and we can do a screen sharing session :)

jinjagit commented:
> Not following, but let me know if you're stuck and we can do a screen sharing session :)

@brentguf Thanks. So far so good. All green with refactored ProjectValueSelector, and manual testing looks good. Starting on TopicValueSelector stuff now.

jinjagit committed:
TopicValueSelector: interface, not type

jinjagit committed:
TopicValueSelector: refactor to functional component

jinjagit committed:
TopicValueSelector: resource hook, not resource

jinjagit committed:
TopicValueSelector: useLocalize

jinjagit committed:
TopicValueSelector: more detailed prop

jinjagit committed:
TopicValueSelector: export props (build error)

jinjagit commented:
@brentguf Please review my refactoring of `ProjectValueSelector.tsx` & `TopicValueSelector.tsx`.  Thanks!

Each is slightly different from your `IdeaStatusvalueSelector.tsx` refactor.  
Most notably we are still using the `getProjects`, not a hook, as no such hook is defined (maybe we should open a new ticket to not lose this?).

brentguf commented:
> @brentguf Please review my refactoring of ProjectValueSelector.tsx & TopicValueSelector.tsx. Thanks!
> 
> Each is slightly different from your IdeaStatusvalueSelector.tsx refactor.
> Most notably we are still using the getProjects, not a hook, as no such hook is defined (maybe we should open a new ticket to not lose this?).

Looks good! No need to log a ticket for useProjects, but you could log it as technical debt (on the TEC board) if you want.

I saw there are still 3 other class components. Can you refactor these to functional ones as well?

jinjagit committed:
Refactor IdeaStatusValuesSelector

Interface, not type
Refactor to functional component
Resource hook, not functional component
useLocalize
More detailed prop
Export props (build error)

jinjagit committed:
Fix: onChange(optionIds[0]); not (optionsIds)

This avoids Argument of type 'any[]' is not assignable to parameter
 of type 'string' error, but will this work correctly?

jinjagit commented:
@brentguf Hi, I’m trying to solve an `Argument of type 'any[]' is not assignable to parameter of type 'string'.` error, caused by `onChange(optionIds)`, in:
```
const handleOnChange = (options: IOption[]) => {
    const optionIds = options.map((o) => o.value);
    onChange(optionIds);
  };
```
in `IdeaStatusValuesSelector.ts`
and have seen this approach, `onChange(optionIds[0])`, used elsewhere. But won’t this always select the first item in the menu?  
  
Testing manually, this approach appears to give a blank dropdown menu, so probably not the right approach!  
  
Using `onChange(optionIds[value]);` means I _can_ now select items from the drop-down menu, but get a `Some conditions are incomplete` error in the smart-group creation form.

brentguf committed:
Solve onChange type problem

jinjagit commented:
Thanks, I had just worked out we needed `onChange: (value: string[]) => void;` too! :+1:

jinjagit committed:
Refactor ProjectValuesSelector

Interface, not type
Refactor to functional component
useLocalize
More detailed prop
Export props (build error)

jinjagit committed:
Refactor TopicValuesSelector

Interface, not type
Refactor to functional component
Resource hook, not functional component
useLocalize
More detailed prop
Export props (build error)

jinjagit commented:
@brentguf Remaining 3 class components refactored. Manual testing appears to show all multi-value predicates working.  
  
I created a TEC ticket: [Create useProjects hook](https://citizenlab.atlassian.net/browse/TEC-78)

brentguf reviewed (approved):
@jinjagit Looks good! 👍
